### ISSUE TITLE
Possibility for deterministic data augmentation in `Dataset`

### ISSUE THREAD

**Issue by prclibo** [pytorch/ignite#929]
## ❓ Questions/Help/Support

Hi I noticed that random augmentation happening inside a `Dataset` is not yet deterministic if resuming from the middle of an epoch. So I would like to ask if this could be possible at the current design of ignite and pytorch.

A short dig into the question results me several difficulties:

1. Random seed is updated per epoch in ignite. So when resuming training at iter 0 and iter 100, I will get different random augmentation at a same step, e.g. iter 101.

2. `DataLoader` has its own seeding strategy for workers. It seems difficult for workers to be deterministically assigned an index and a seed.

It would be great if anyone have experience on this and can share some thoughts.

**Comment by sdesrozis** [comment]
Thanks for your question. I don't have input on that atm. 

@vfdev-5 ?

**Comment by vfdev-5** [comment]
@prclibo this is a good question, thanks !

Yes, this is a tough task to resume from an interation with all possible randomness that could be encountered, e.g dataflow (random sampling, data augs), model (dropout), random handlers etc...

Maybe, ultimately deterministic run can be done with synchronizing rng states on each iteration what leads to large overhead. And even in this case data prefetching with DataLoader can lead some difficulties (on iteration N, RNG is sync for Nth iteration and dataloader prefetched data with dataaugs for next m iterations. If we would like to resume from N + m/2 data wont be the same...)

Current implementation of ignite's engine synchronizes rng state on every epoch and controls batch indexes if dataloader is torch dataloader. This implementation seems very restrictive and will be replaced in v0.4.0, #895. In this PR we will provide some bricks to make trainer deterministic and syncronize RNG states from the checkpoint. In some cases, this can lead to "exactly" the same data and model gradients (without forgetting to set cudnn to deterministic).

**Comment by prclibo** [comment]
@vfdev-5 Thanks for the reply and it is nice to know there are some PRs dealing with similar questions.

I agree that data prefetching will easily break the rng state synchronization strategy. Not to mention that things are messier when DataLoader is executed multi-threaded. Will you have plan to deal with this prefetching or multi-thread case? My naive idea is to bind each data index with a corresponding seed, which can probably be done in a Sampler. DataLoader can then synchronize rng state / setup seed before calling `next(self._dataloader_iter)`. But this might probably require to change code of `DataLoader`, especially when dealing with multi-thread worker loop.

**Comment by vfdev-5** [comment]
@prclibo Ignite deals with data sampling to make it deterministic in the following way if dataloader is torch `DataLoader` of cause:
1 replace batch sampler by custom batch sampler
2 get all sample indices for 1 run of dataloader before restarting
3 register them in custom batch sampler and use during the run
4 on dataloader restart, repeat from 2

**Comment by prclibo** [comment]
@vfdev-5 Thanks for the explanation.

**Comment by vfdev-5** [comment]
@prclibo I close the issue as solved. Feel free to reopen if needed.
